src/CloudNative.CloudEvents/CloudEventContent.cs

```diff
@@ -75,32 +78,40 @@ protected override bool TryComputeLength(out long length)
             return inner.InnerTryComputeLength(out length);
         }
 
-        void MapHeaders(CloudEvent cloudEvent)
+        void MapHeaders(CloudEvent cloudEvent, bool includeDataContentType)
         {
+            string specVersionAttributeName = CloudEventAttributes.DataAttributeName(cloudEvent.SpecVersion);
+            string dataContentTypeAttributeName = CloudEventAttributes.DataContentTypeAttributeName(cloudEvent.SpecVersion);
+
             foreach (var attribute in cloudEvent.GetAttributes())
             {
-                if (!(attribute.Key.Equals(CloudEventAttributes.DataAttributeName(cloudEvent.SpecVersion)) ||
-                      attribute.Key.Equals(CloudEventAttributes.DataContentTypeAttributeName(cloudEvent.SpecVersion))))
+                string key = attribute.Key;
+                string headerName = "ce-" + key;
+                object value = attribute.Value;
+
+                // Never map the spec attribute to a header
+                if (key == specVersionAttributeName)
                 {
-                    if (attribute.Value is string)
-                    {
-                        Headers.Add("ce-" + attribute.Key, WebUtility.UrlEncode(attribute.Value.ToString()));
-                    }
-                    else if (attribute.Value is DateTime)
-                    {
-                        Headers.Add("ce-" + attribute.Key, ((DateTime)attribute.Value).ToString("u"));
-                    }
-                    else if (attribute.Value is Uri || attribute.Value is int)
-                    {
-                        Headers.Add("ce-" + attribute.Key, attribute.Value.ToString());
-                    }
-                    else
+                    continue;
+                }
+                // Only map the data content type attribute to a header if we've been asked to
+                else if (key == dataContentTypeAttributeName && !includeDataContentType)
+                {
+                    continue;
+                }
+                else
+                {
+                    string headerValue = attribute.Value switch
                     {
-                        Headers.Add("ce-" + attribute.Key,
-                            WebUtility.UrlEncode(
-                                Encoding.UTF8.GetString(jsonFormatter.EncodeAttribute(cloudEvent.SpecVersion, attribute.Key, attribute.Value,
-                                    cloudEvent.Extensions.Values))));
-                    }
+                        string text => WebUtility.UrlEncode(text),
+                        ContentType contentType => contentType.ToString(),
+                        DateTime dt => XmlConvert.ToString(dt, XmlDateTimeSerializationMode.Utc),
```

Comment:
Out of curiosity, why do we use XmlConvert here?

Reply:
Because it performs exactly the formatting we need it to, and that's the simplest way of doing so. (It's been a while since I looked at this, but I don't think there's a very simple date/time format that does what we need.)